artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/connect/mirror/AMQPMirrorControllerTarget.java

```diff
@@ -267,102 +310,143 @@ public void deleteAddress(AddressInfo addressInfo) throws Exception {
+      }
+      MessageReference reference = targetQueue.removeWithSuppliedID(nodeID, messageID, referenceIDSupplier);
+      if (reference == null && retry) {
+         if (logger.isDebugEnabled()) {
+            logger.debug("Retrying Reference not found on messageID=" + messageID + " nodeID=" + nodeID);
+         }
+         targetQueue.flushOnIntermediate(() -> {
+            recoverContext();
+            performAck(nodeID, messageID, targetQueue, ackMessageOperation, false);
+         });
+         return;
+      }
+      if (reference != null) {
+         if (logger.isTraceEnabled()) {
+            logger.trace("Post ack Server " + server + " worked well for messageID=" + messageID + " nodeID=" + nodeID);
+         }
+         try {
+            targetQueue.acknowledge(reference);
+            OperationContextImpl.getContext().executeOnCompletion(ackMessageOperation);
+         } catch (Exception e) {
+            logger.warn(e.getMessage(), e);
+         }
+      } else {
+         if (logger.isDebugEnabled()) {
+            logger.debug("Post ack Server " + server + " could not find messageID = " + messageID +
+                            " representing nodeID=" + nodeID);
          }
       }
 
    }
 
-   private void sendMessage(AMQPMessage message) throws Exception {
+   private boolean sendMessage(AMQPMessage message, ACKMessageOperation messageCompletionAck) throws Exception {
+
       if (message.getMessageID() <= 0) {
          message.setMessageID(server.getStorageManager().generateID());
       }
 
-      Long internalID = (Long) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_ID);
+      String internalMirrorID = (String)AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, BROKER_ID);
+      if (internalMirrorID == null) {
+         internalMirrorID = getRemoteMirrorId(); // not pasisng the ID means the data was generated on the remote broker
+      }
+      Long internalIDLong = (Long) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_ID);
       String internalAddress = (String) AMQPMessageBrokerAccessor.getDeliveryAnnotationProperty(message, INTERNAL_DESTINATION);
 
-      if (internalID != null) {
+      long internalID = 0;
+
+      if (internalIDLong != null) {
+         internalID = internalIDLong;
+      }
+
+
+      if (logger.isTraceEnabled()) {
+         logger.trace("sendMessage on server " + server + " for message " + message +
+                         " with internalID = " + internalIDLong + " mirror id " + internalMirrorID);
+      }
+
+      final TransactionImpl transaction = new MirrorTransaction(server.getStorageManager());
+      transaction.addOperation(messageCompletionAck);
+
+      routingContext.setDuplicateDetection(false); // we do our own duplicate detection here
+
+      if (internalID != 0) {
+         byte[] duplicateIDBytes = ByteUtil.longToBytes(internalID);
```

Comment:
This value is now only using the long id, which every broker generates the same sequence and so can overlap. Seems like it needs the server id to be considered too if its desired to support e.g chained mirrors etc as you have mentioned?

Reply:
I was actually thinking about this...  yeah.. I need to implement that.

Reply:
fixed

Reply:
Actually I will try to refactor the duplicate detection here.
instead of merging the duplicateID for every message received, I will try to play with separate duplicate detection tables. one for each server.... I will be doing that it if it's not too hard.